### DISCUSSION TITLE
Memory Synchronization Failure

### DISCUSSION THREAD

**Discussion by kentslaney** [gfx-rs/wgpu#6629] [Q&A] [answered]
Context:

The [wgpu project that I'm working on](https://github.com/kentslaney/wgpu_nnd/tree/e13c89d9604c52304206be5f3eccfe963704292f) seems to produce a race condition which causes different threads to see different values even after a `storageBarrier`. On a system with Python in its executable paths, the problem should be reproducible by running `make`, which causes a log info line labelled `Candidates` to give an array which should have uniform values. On my system, with a Vulkan `1.3.295` back end, it produces inconsistent values.

Apologies that the example isn't more minimal. The relevant code is essentially just [this line](https://github.com/kentslaney/wgpu_nnd/blob/e13c89d9604c52304206be5f3eccfe963704292f/src/shader.wgsl#L546) which checks the same location and writes to different locations in the same storage buffer on different threads. The value should be `1`, but is a mix of `0`, `1`, and `3` when I run it.

My questions are:
- **Are atomic operations needed for writing different locations in the same buffer?** (I've been assuming not)
- Is there another reason this may not represent unexpected behavior?
- Can other people reproduce the problem?
- Are there other mistakes that I've missed?
- Assuming that the infrastructure is working as expected, what can I do to solve my problem?

Thanks in advance.

**Comment by cwfitzgerald** [comment, marked as answer]
This is a petty common misconception about storageBarrier(), it only applies within the workgroup. That is, that ensures the entire workgroup will see all of the workgroups writes to global memory. If you need to synchronize the entire dispatch, you need to use two dispatches.

**Reply by kentslaney** [reply]
That would explain the behavior, thanks

**Reply by kentslaney** [reply]
I think this means #4450 affects me now though
